Applications/Spire/Include/Spire/Ui/NumericFilterPanel.hpp

```diff
@@ -0,0 +1,274 @@
+        m_default_max_value(std::move(default_max_value)) {
+    if(m_default_min_value && m_default_max_value &&
+        *m_default_min_value > *m_default_max_value) {
+      std::swap(m_default_min_value, m_default_max_value);
+    }
+    auto layout = new QVBoxLayout(this);
+    layout->setSpacing(0);
+    layout->setContentsMargins({});
+    auto min_input = make_input_field(m_min_model);
+    auto min_layout = make_row_layout(tr("Min"), min_input);
+    layout->addLayout(min_layout);
+    layout->addSpacing(scale_height(10));
+    auto max_input = make_input_field(m_max_model);
+    auto max_layout = make_row_layout(tr("Max"), max_input);
+    layout->addLayout(max_layout);
+    m_filter_panel = new FilterPanel(std::move(title), this, parent);
+    m_filter_panel->connect_reset_signal([=] {
+      reset();
+    });
+    min_input->connect_submit_signal([=] (const auto& value) {
+      if(m_max_value && value && *value > *m_max_value) {
+        m_max_model->set_current(value);
+      }
+      if(m_min_value != value) {
+        m_filter_signal(value, m_max_value);
+      }
+      m_min_value = value;
+    });
+    max_input->connect_submit_signal([=] (const auto& value) {
+      if(m_min_value && value && *value < *m_min_value) {
+        m_min_model->set_current(value);
+      }
+      if(m_max_value != value) {
+        m_filter_signal(m_min_value, value);
+      }
+      m_max_value = value;
+    });
+    setFocusProxy(min_input);
+  }
+
+  template<typename T>
+  const std::shared_ptr<typename NumericFilterPanel<T>::OptionalNumericModel>&
+      NumericFilterPanel<T>::get_min_model() const {
+    return m_min_model;
+  }
+
+  template<typename T>
+  const std::shared_ptr<typename NumericFilterPanel<T>::OptionalNumericModel>&
+      NumericFilterPanel<T>::get_max_model() const {
+    return m_max_model;
+  }
+
+  template<typename T>
+  boost::optional<typename NumericFilterPanel<T>::Type>
+      NumericFilterPanel<T>::get_default_min_value() const {
+    return m_default_min_value;
+  }
+
+  template<typename T>
+  boost::optional<typename NumericFilterPanel<T>::Type>
+      NumericFilterPanel<T>::get_default_max_value() const {
+    return m_default_max_value;
+  }
+
+  template<typename T>
+  void NumericFilterPanel<T>::set_default_value(
+      const boost::optional<Type>& default_min,
+      const boost::optional<Type>& default_max) {
+    m_default_min = default_min;
+    m_default_max = default_max;
+  }
+
+  template<typename T>
+  boost::signals2::connection NumericFilterPanel<T>::connect_filter_signal(
+      typename const FilterSignal::slot_type& slot) const {
+    return m_filter_signal.connect(slot);
+  }
+
+  template<typename T>
+  bool NumericFilterPanel<T>::event(QEvent* event) {
+    if(event->type() == QEvent::ShowToParent) {
+      reset();
+      m_filter_panel->show();
+    }
+    return QWidget::event(event);
+  }
+
+  template<typename T>
+  typename NumericFilterPanel<T>::NumericBox*
```

Comment:
This is declared to be an instance method, but it does not make use of ```this```. Use ```static``` for methods that are not tied to any instance.